### DISCUSSION TITLE
Range Extender POC

### DISCUSSION THREAD

**Discussion by sillyfrog** [arendst/Tasmota#12421] [Ideas]
I have added (early) support for adding a "Range extender" feature to Tasmota. You can see a proof of concept here: https://github.com/sillyfrog/Tasmota/tree/sillyfrog-rangeextender

This allows you to add a new WiFi AP to a Tasmota device that will route the traffic through to the network the Tasmota is connected to. I have used this for a little while now and it's going well (I have just updated for the latest v9.5 code).

Please note: this branch includes a `platformio_tasmota_cenv.ini` and `tasmota/user_config_override.h` file so you can clone and build easily.

In `tasmota/user_config_override.h` are the current config options (questions on this below) - including the AP SSID/Password, IP address, netmaks and if you want to enable NAT/NAPT.

The NAPT support uses about 3.2k of the heap, but makes using the range extender feature really easy, just join the network created above, and your devices will have access to the rest of the network. However you won't be able to access the devices as their are behind the NAPT. With out NAPT you need to setup routes back to the network behind the Tasmota device on your core router - but you can then access the devices behind by IP address (eg: get to the web interface). If using MQTT, NAPT is probably the easiest way to go.

Some questions about implementing this:
- At the moment the changes are directly in `support_wifi.ino` (inside `#define`s) - is there an API hook I could use to move this to a driver file so it gets called when WiFi comes up?
- At the moment the settings are all hard coded, and I think it would be ideal if they were to me be moved to config options. Is there a best practice example for how to add config options? Should this be a `SetOption` or have custom commands? I looked at the source for `SSID` etc, but I didn't find it obvious the best way to do things.

Any thoughts/feedback appreciated - if it's something others may see of value I'd be keen to get it polished and added to Tasmota (as an option for custom builds).

Thanks!

**Comment by joba-1** [comment]
Thanks for providing this, can get very useful to me.

As you noted I implemented a client list command and my reason for this was some rgx connection issues I wanted to debug.
Unfortunately they are not solved by this, just easier to see:

* The Rgx AP is disappearing entirely after some time (so far ~0,5-6 hours, not sure there is a pattern to this yet). 
* Rgx clients get disconnected although within stable, close range and can reconnect easily.
* The tasmota wifi scan seems to not cause this, I've seen clients remain the same before and after a scan.

Other tasmota functionality (e.g. STA wifi including mqtt, syslog and web OTA) seems to be unaffected by this.

Nothing that would help me in the serial log level 4 at the time a client goes away:
```
11:29:12.891 SRC: MQTT
11:29:12.892 CMD: Grp 0, Cmd 'RGXCLIENTS', Idx 1, Len 0, Pld -99, Data ''
11:29:12.904 MQT: stat/nat2/RESULT = {"RgxClients":{"4c:eb:d6:77:f7:60":{"IPAddress":"192.168.4.2","RSSI":-85},"ac:5f:3e:50:65:bb":{"IPAddress":"192.168.4.3","RSSI":-64}}}
11:29:20.361 WIF: Checking connection...
11:29:40.378 WIF: Checking connection...
11:30:00.431 WIF: Checking connection...
11:30:10.046 BRY: GC from 4404 to 3702 bytes, objects freed 5/40 (in 1 ms) - slots from 52/122 to 42/122
11:30:10.055 MQT: tele/nat2/STATE = {"Time":"2022-11-14T11:30:10","Uptime":"0T00:30:16","UptimeSec":1816,"Heap":126,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":40},"Wifi":{"AP":1,"SSId":"job.fritz.ssid","BSSId":"28:D1:27:4E:56:73","Channel":1,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:24"}}
11:30:12.837 SRC: MQTT
11:30:12.839 CMD: Grp 0, Cmd 'RGXCLIENTS', Idx 1, Len 0, Pld -99, Data ''
11:30:12.860 MQT: stat/nat2/RESULT = {"RgxClients":{"4c:eb:d6:77:f7:60":{"IPAddress":"192.168.4.2","RSSI":-84}}}
```

maybe there is some mandatory communication between client and AP that is not implemented?

**Comment by joba-1** [comment]
FYI: I also added an RgxPort command now. So you can now access devices behind the NAT. 
After some trouble it is also rock solid for me now. Not a single disconnect, vanishing AP or other problem since 2 days+ :)

**Comment by UBWH** [comment]
I also have found the RangeExtender super useful.

But - hit a snag.
I was hoping to use the AP side of RGX as a standalone AP on Tasmota all the time. I did lots of testing and all worked the way I wanted when the Tasmota WiFi client connected to my local AP.

Why do I want a Tasmota AP? This could allow access to the Tasmota MAIN menu even if the WiFi interface has no client IP address. Great for viewing Tasmota webpages in a remote location.

Discovered the if I take the Tasmota device away from my local environment - where the WiFi client fails to connect to my local AP, then the RGX AP fails to start. This makes a lot of sense as the purpose of RGX is to pass packets between the RGX AP and the WiFi client.

But - it has messed up my plans to have a Tasmota device that has a built in AP running all the time.

I am good at C## coding - so had a look at the RGX driver [xdrv_58_range_extender.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_58_range_extender.ino). I commented out some of the checks like `WiFi.getMode() == WIFI_AP_STA` (and built a custom *.bin) in the hope RGX will still run, but no luck.

My question is:
Any advice you can offer so that I can keep the AP running even when the WiFi client is not connected?

**Reply by sillyfrog** [reply]
Long time since I did this, but I _think_ if you do something like this, it _may_ get it going (this is a guess, untested):
```
diff --git a/tasmota/tasmota_xdrv_driver/xdrv_58_range_extender.ino b/tasmota/tasmota_xdrv_driver/xdrv_58_range_extender.ino
index 3868dde46..2f700bb40 100644
--- a/tasmota/tasmota_xdrv_driver/xdrv_58_range_extender.ino
+++ b/tasmota/tasmota_xdrv_driver/xdrv_58_range_extender.ino
@@ -492,7 +492,7 @@ bool Xdrv58(uint32_t function)
       break;
     case FUNC_EVERY_SECOND:
       // AddLog(LOG_LEVEL_INFO, PSTR("RGX: XXX DEBUG: Wifi.status: %d, WiFi.getMode(): %d, RgxSettings.status: %d, link_count: %d"), Wifi.status, WiFi.getMode(), RgxSettings.status, Wifi.link_count);
-      if (RgxSettings.status == RGX_NOT_CONFIGURED && Wifi.status == WL_CONNECTED)
+      if (RgxSettings.status == RGX_NOT_CONFIGURED)
       {
         // Setup only if WiFi in STA only mode
         if (WiFi.getMode() == WIFI_STA)
```
I however have a vague memory that it relies on some other setup from the core of Tasmota to get the WiFi started - but can't hurt giving this a try and seeing if it works.

**Reply by UBWH** [reply]
Thanks for the quick reply.

Yeah - already tried that - and WiFi AP did not start.

I think I will find the code that turns off the management AP (192.168.4.1) after 3 mins, and keep it on. I find that does give me access (for 3 mins) to the web pages if WiFi client is configured ... but not connected.

**Comment by fdaubercy** [comment]
I'm digging up this topic.
I want to be able to access the web pages of a tasmota module not connected to the Wi-Fi network through the RangeExtender access point.
However, this access point won't start if the Wi-Fi access point isn't connected to the local network.
I have disabled the the wifi_manager running with the '#define WIFI_CONFIG_TOOL WIFI_WAIT' parameter
**How can I keep the RangeExtender access point running even when the Wi-Fi client isn't connected to the local network?**
My goal here is to be able to connect to the tasmota module through its access point and view its web page.

**Reply by sfromis** [reply]
As the [docs page](https://tasmota.github.io/docs/Range-Extender/) for the feature starts:
> This feature allows Tasmota to publish an AP with a dedicated SSID that is then routed to the core WiFi network.

And later:
> this feature allows you to turn it into a routing WiFi node.

What you ask about is simply outside of the purpose of being a "range **extender**", and basically equivalent to what was already asked in February 2024

**Reply by pkkrusty** [reply]
If you're just looking for offline access, there is a hack to let you use the AP all the time. Caveat that you will not have correct timing. I use it successfully in a bunch of different areas. (Especially useful when paired with DS3231 module)

See discussion #19009
